docs/development/review.md

```diff
@@ -75,51 +129,52 @@ kindly,
-
-
-A fourth type he sometimes uses is the reaction, which is not actionable, but might help to express emotions such as thanks.
-
-He further suggests to always keep things as clear as possible by using the following pattern:
-
->[type]: [actionable request] 
-
->[rationale or discussion]
-
-Whereas:
-- type: one of Question, Suggestion or Changes Request.
-- actionable request: is a short clear question or statement, which makes clear how the reviewee
-should respond.
-- rational or discussion: Possible further explanations of the reasoning behind the request.
-
-An example could be: 
->Suggestion: use XYZ. 
-
->You may have come across this already, but we might be able to use XYZ for this because blah blah blah etc.
-
-We adapted this slightly by bundling questions, suggestions and change requests in grouped listings and placing
-the reactions up front.
-
-## Inline Comments
-Github makes it very easy to leave inline comments. However, in the past, we observed, that large
-numbers of inline comments make a review hard to read and understand. We therefore propose
-only the leave inline comments for trivial directly actionable suggestions and change request, that
-do not need an rational (E.g. *Change Request: typo in word actionalbe*).
-
-Do not leave any reactions in inline comments. They tend to crowd the discussion without giving
-the possibility of being resolved (seens not actionable). Move them to the summary.
+Please commit the following file as part of the review to the root of the component.
 
+#### If the file already exists:
+* Please update the sections 'Maintenance' and 'Statistics'.
+* Please add the sections 'Criteria based Review for [ilias_version]' und 'Qualitative Review for [ilias_version]' **right below** the section 'Statistics'
 
+#### Review.md
 
+```
+# Review of `[name_of_component]` by [name_of_reviewer] at [date_of_review]
+
+## Maintenance
+* Maintainers/Coordinators:
+* Committers in last 365 days:
+
+## Statistics
+* Lines of Code: 
+* Nr of Tests: 
+* LoCs changed in last 365 days:
+* Nr of Commits in last 365 days: 
+* Nr of Open Mantis Issues: 
+* Mantis Issues resolved in last 365 days: 
+* Nr of Votes in last 365 days:
+* Nr of Votes for currently open issues:  
+
+## Criteria based Review for [ilias_version]
+
+* [ ] `[name_of_component]`and its administration can be used in devmode with PHP 8 and 8.1 without producing PHP-specific errors
+* [ ] The code style is PSR-12 + snake case compliant
+* Globals:
+	* [ ] The code contains no $_GET / $_POST / $_REQUEST / $_SESSION
+	* [ ] Access to $_GET-parameters happens through the request-wrapper and only in constructors
+	* [ ] Direct access to $_POST parameters through the request-wrapper is only used exceptionally.
+* [ ] There are at least as many unit tests as in the last ILIAS Version and all tests pass
+* [ ] All external libraries are up to date and compatible with PHP 8 and 8.1
+* [ ] There are no serious PhpStorm Code Inspection / PHPStan issues left
```

Comment:
Ich bin dagegenm die PHPStorm-Inspection hier zu verstetigen. Können wir komplett auf PHPStan umsteigen? Ich glaube, es müsste dann aber eine Stufe für das Tool angegeben werden.

Reply:
Ich bin dafür, PHPStorm und die Inspection zu fördern und auch teilweise zu verstetigen. Mir ist klar, dass wir uns damit teilweise an ein Produkt binden (das bspw. gerade seine Preise erhöht). PHPStan kann gerne als Alternative oder Fallback erwähnt werden, aber zwecks Effizienz, Wartbarkeit von Profilen und Vergleichbarkeit von Resultaten stehe ich hinter dieser Quasi-Standard-IDE. Ich sehe fast täglich Beispiele im ILIAS-Code, die uns nicht "passieren" würden, wäre die Nutzung deer IDE verpflichtend. Ich weiss da habe ich eine starke Meinung zu und @klees und ich werden uns diesbezüglich nie finden ;-)

Reply:
+1 @chfsx .

Reply:
PHPStorm und  dessen Inspections zu fördern ist ja was anderes als sie zu fordern =)

Wenn wir PHPStorm in irgendeiner Form dauerhaft verpflichtend machen (was dies hier ja wäre) schließen wir folgende Gruppen aus der Entwicklung aus:
* Leute, die sich die Lizenz preislich nicht leisten können oder wollen
* Leute, die keine Geschäfte mit JetBrains machen dürfen (Russland-Sanktionen...)

Insbesondere den ersten Punkt finde ich sehr relevant: PHPStorm wurde in der ILIAS-Community damals überhaupt nur so breit eingeführt, weil kostenlose Lizenzen von JetBrains angeboten wurden. Das wurde ja derweil wiederrufen und der Preis der Lizenzen wird auch angehoben(?) ...

Daneben sprechen noch andere Gründe dagegen, PHPStorm für diesen Kontext zu fordern:
* Die Inspection lässt sich nicht in einem CI-Prozess einbinden, d.h. bzgl. Code Quality wären wir bei CI dann an einem Dead End.
* Die Inspection basiert auf Einstellungen, die teilweise lokal geändert werden. Was ja sogar wünschenswert ist (eigener Stil, Präferenzen, Qualitätsansprüche, ...), aber Reproduzier- und Vergleichbarkeit stört. Nämliches Problem sehe ich immer wieder im Kontext von Coding-Style in PRs.
* Es gibt keine klare oder nachvollziehbare Aussage oder Doku darüber, was die PHPStorm-Inspections genau fordern (ist mir zumindestens nicht bekannt). Das macht den Inhalt der Inspection schwer besprech-, verhandel- und kommunizierbar.

Ich kann also nicht nachvollziehen, wie deine Aussage zu Effizienz, Wartbarkeit und Vergleichbarkeit zustande kommen.

Mein Argumente hier ist also eines, das ohne ein generelles pro/contra IDE auskommt. Deren Nutzen sehe ich generell, aber auch nicht unbeschränkt. Ich stimme zu, dass es viel Code gibt, der mit IDE nicht passiert wäre. Andersrum sehe ich aber auch Code mit vielen Komplexitäten, der ohne IDE besser oder klarer designed sein müsste und der ohne Unterstützung nicht gut zu verstehen oder zu bearbeiten ist. Oder anders: Code ist ja nicht deshalb gut, weil die IDE keine Warnings mehr schmeißt.

Ich finde also, dass wir gerne weiter PHPStorm nahelegen sowie Konfiguration und Handreichungen bereitstellen können und finde das auch gut. Aber darum geht es hier nicht.

Reply:
Ich verstehe alle deine Argumente und bin da auch voll bei dir. Ich persönlich bin einfach irgendwann abgehängt oder nicht motiviert etwas zu tun, wenn ich durch viel Gefrickel (und das ist leider für mich persönlich bspw. beim CodeStyle bereits der Fall, meine Meinung, müssen wir nicht diskutieren) meine Performance in den Keller geht. Ich bin aber voll dabei, beide Varianten zu akzeptieren, bspw:
- Review auf Basis von PHStorm
- Review auf Basis von anderen Tools
Reviews auf Basis von einem Blumenstrauss von Tools würde ich selber bspw. wohl eher selten durchführen resp. anbieten.

Reply:
Dabei müssten wir aber auch akzeptieren, dass die beiden Formen von Reviews zu unterschiedlichen Resultaten führen wird, so vermute ich.

> Andersrum sehe ich aber auch Code mit vielen Komplexitäten, der ohne IDE besser oder klarer designed sein müsste und der ohne Unterstützung nicht gut zu verstehen oder zu bearbeiten ist.

Das sehe ich nicht so, Beispiele würden da aber vielleicht helfen. Die Komplexität von Code nimmt nicht zu, weil eine IDE im Spiel ist, sondern hängt damit zusammen, wie der Code designed wird. Da kommen m.E. andere Dinge zu tragen wie der eigene Bezug zu "convention vs. declaration"

Reply:
Das hier scheint der einzige wirkliche Diskussionspunkt zu sein.

Ich glaube aber, bei näherer Überlegung, dass es nur zwei Lösungen gibt:

- Entweder wir fassen diesen Punkt Tool agnostisch, d.h. wir schreiben unsere Kritierien auf und schlagen PHPStorm und PHPStan mit den entsprechenden Einstellungen als Analysetools vor.
- Oder wir nehmen nur PHPStan.

Der Grund dafür ist einfach:
- Wenn da steht, dass mein Code noch 10 Fehler in den Code-Inspections von PHPStorm hat, dann sage ich: "Nice und was nun?". Als nicht PHPStorm-Nutzer kann ich auf keine Art, ausser in mühsamster Pfrimelarbeit, herausfinden, was das nun genau heisst.
- Nur wenn ich die Kriterien habe oder mir ein freies (wie in Freibier) Werkzeug zur Verfügung steht, kann ich die Rückmeldung nachvollziehen.

Ich habe versucht die Kriterien von PHPStorm zusammenzutragen. Wenn ich das Inspection-Profile auf "enabled='true'" greppe, dann finde ich die folgenden eingeschalteten Inspections:
```
<inspection_tool class="PhpCSFixerValidationInspection" enabled="true" level="WEAK WARNING" enabled_by_default="true">
<inspection_tool class="PhpDivisionByZeroInspection" enabled="true" level="WARNING" enabled_by_default="true" />
<inspection_tool class="PhpOverridingMethodVisibilityInspection" enabled="true" level="WEAK WARNING" enabled_by_default="true" />
<inspection_tool class="PhpUndefinedCallbackInspection" enabled="true" level="WARNING" enabled_by_default="true" />
<inspection_tool class="PhpUsageOfSilenceOperatorInspection" enabled="true" level="WEAK WARNING" enabled_by_default="true" />
<inspection_tool class="PhpVariableVariableInspection" enabled="true" level="WEAK WARNING" enabled_by_default="true" />
<inspection_tool class="SSBasedInspection" enabled="true" level="WARNING" enabled_by_default="true" />
```
Wenn ich auf ihrer Seite suche, dann finde ich dies: https://www.jetbrains.com/help/phpstorm/code-inspections-in-php.html

Ich bringe die beiden Dinge aber nicht zusammen ;-), vielleicht hat mir jemand ja einen Tipp.